lib/fastlyPromise.js

```diff
@@ -145,4 +149,170 @@ FastlyPromise.prototype.purgeKey = function purgeKey(serviceId, surrogateKey, so
+ *
+ * @param serviceId
+ * @param configVersionNumber
+ * @param vclName
+ */
+FastlyPromise.prototype.getVcl = function getVcl(serviceId, configVersionNumber, vclName) {
+
+  return this.request('GET', this.endpoint + '/service/' + serviceId + '/version/' + configVersionNumber + '/vcl/' + vclName);
+
+}
+
+/**
+ *
+ * Get the currently active VCL for a service.
+ *
+ * @param {string} serviceId
+ * @param {string} [configVersionNumber] - If no config version number is passed, defaults to active config version.
+ */
+FastlyPromise.prototype.getActiveVcl = function getActiveVcl(serviceId, configVersionNumber) {
+  var self = this;
+
+  //Set default config version number to active version if not explicitly defined.
+  var configVersionNumberPromise = configVersionNumber ? Promise.resolve(configVersionNumber) : self.getActiveConfigVersion(serviceId).then(function(version){ return version.number });
+
+  return configVersionNumberPromise
+    .then(function(configVersionNumber){
+    return self.getAllVcl(serviceId, configVersionNumber);
+  })
+    .then(function(vclList){
+
+      return vclList.find(function(vcl){
+        return vcl.main;
+      });
+
+    });
+
+}
+
+/**
+ *
+ * Upload new VCL to a service. Throws error if VCL with same name already exists.
+ *
+ * @param {string} serviceId
+ * @param {string} configVersionNumber - Service config version number.
+ * @param {string} vclName - Name of vcl file.
+ * @param {string} vclContent - The VCL content to be uploaded.
+ * @param {boolean} [setVclToMain=false] - Set the uploaded vcl to be the service's "main" vcl.
+ */
+FastlyPromise.prototype.uploadNewVcl = function uploadNewVcl(serviceId, configVersionNumber, vclName, vclContent, setVclToMain) {
+  var self = this;
+
+  setVclToMain = setVclToMain || false;
+
+  var validNamePromise = self.getVcl(serviceId, configVersionNumber, vclName)
+    .then(function(response){
+
+      return false; //Invalid if vcl with this name already exists
+
+    })
+    .catch(function(errorResponse){
+
+      //If response code is 404 then new vcl name is valid.
+      return (errorResponse.statusCode == 404) ? true : false;
+
+    });
+
+  return validNamePromise
+    .then(function(validName){
+
+      if (!validName){
+        throw new Error('Error: ' + vclName + ' VCL file already exists.');
+      }
+      else{
+
+        //Upload the VCL
+        return self.request('POST', self.endpoint + '/service/' + serviceId + '/version/' + configVersionNumber + '/vcl', {
+          form: {
+            name: vclName,
+            content: vclContent
+          }
+        });
+
+      }
+
+    })
+    .then(function(vclResponse){
+
+      //Set VCL to "main" if flagged.
+      if (setVclToMain){
+        return self.request('PUT', self.endpoint + '/service/' + serviceId + '/version/' + configVersionNumber + '/vcl/' + vclName + '/main');
+      }
+      else{
+        return vclResponse;
+      }
+
+    });
+
+}
+
```

Comment:
Are there any VCL-related endpoints that we haven't implemented? If so, should we stub them out so we know what we didn't actually implement?